File: roman-numeral/js/roman.js
const conversions = {
  I: 1,
  IV: 4,
  V: 5,
  IX: 9,
  X: 10,
  XL: 40,
  L: 50,
  XC: 90,
  C: 100,
  D: 500,
  CM: 900,
  M: 1000,
}

export const fromRoman = (roman) => {
  // return addArabic(roman, 0)
  return addArabicIterative(roman)
}

const addArabic = (roman, total) => {
  if (roman.length === 0) return total
  const head = getNextLongestMatchingKey(roman)
  const rest = roman.replace(head, '')
  return addArabic(rest, total + conversions[head])
}

const addArabicIterative = (roman) => {
  let total = 0
  while (roman.length > 0) {
    const head = getNextLongestMatchingKey(roman)
    const rest = roman.replace(head, '')
    total += conversions[head]
    roman = rest
  }
  return total
}

const getNextLongestMatchingKey = (roman) => {
  return roman in conversions
    ? roman
    : getNextLongestMatchingKey(roman.slice(0, roman.length - 1))
}

export const toRoman = (arabic) => {
  // return appendRoman(arabic)
  return appendRomanIterative(arabic)
}

const appendRoman = (arabic) => {
  if (arabic === 0) return ''

  const roman = Object.keys(conversions)
    .reverse()
    .find((r) => conversions[r] <= arabic)
  return roman + appendRoman(arabic - conversions[roman])
}

const appendRomanIterative = (arabic) => {
  let roman = ''

  while (arabic > 0) {
    const nextRoman = Object.keys(conversions)
      .reverse()
      .find((r) => conversions[r] <= arabic)
    arabic -= conversions[nextRoman]
    roman += nextRoman
  }
  return roman
}

File: roman-numeral/ruby/lib/roman.rb
class Roman

  def self.to(arabic)
    return "" if arabic == 0

    num, numeral = @@conversions.find { |entry| entry.first <= arabic }
    numeral + to(arabic - num)
  end

  def self.from(roman, total=0)
    return total if roman == ""

    numeral = get_next_longest_sub_numeral roman
    rest = roman[numeral.length..]
    num, _numeral = @@conversions.find { |entry| entry.last === numeral }
    from rest, total + num
  end

  def self.get_next_longest_sub_numeral(roman)
    _num, numeral = @@conversions.find { |entry| roman == entry.last }
    return numeral if numeral

    self.get_next_longest_sub_numeral(roman[..-2])
  end

  @@conversions = [
    [1000, "M"],
    [900, "CM"],
    [500, "D"],
    [100, "C"],
    [90, "XC"],
    [50, "L"],
    [40, "XL"],
    [10, "X"],
    [9, "IX"],
    [5, "V"],
    [4, "IV"],
    [1, "I"]
  ]

  private_class_method :get_next_longest_sub_numeral

end
